### PR TITLE
Issue #222: Provide hooks that fire before and after a request

### PR DESCRIPTION
**Pull request by pfrenssen**

### CODE REVIEW

File: spec/Behat/MinkExtension/Context/Initializer/MinkAwareInitializerSpec.php [hunk 1]
```diff
@@ -26,8 +34,13 @@ function it_does_nothing_for_basic_contexts(Context $context)
 
     function it_injects_mink_and_parameters_in_mink_aware_contexts(MinkAwareContext $context, $mink)
     {
+        // HookDispatcher() is a final class so it cannot be reflected into a
+        // mocked instance. Let's instantiate a real object.
+        $dispatcher = new HookDispatcher(new HookRepository(new EnvironmentManager()), new CallCenter());
```
Comment (stof) [inline spec/Behat/MinkExtension/Context/Initializer/MinkAwareInitializerSpec.php]: this is wrong. It is not the same instance.

File: src/Behat/MinkExtension/Hook/Scope/BaseRequestScope.php [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+use Behat\Mink\Mink;
+use Behat\Mink\Session;
+use Behat\Testwork\Hook\Scope\HookScope;
+
+/**
+ * Base class for request scopes.
+ *
+ * @author Pieter Frenssen <pieter@frenssen.be>
+ */
+abstract class BaseRequestScope implements RequestScope
```
Comment (stof) [inline src/Behat/MinkExtension/Hook/Scope/BaseRequestScope.php]: there is no concrete implementation of this class

File: src/Behat/MinkExtension/Context/MinkAwareContext.php [hunk 3]
```diff
@@ -33,4 +34,11 @@ public function setMink(Mink $mink);
      * @param array $parameters
      */
     public function setMinkParameters(array $parameters);
+
+    /**
+     * Sets event dispatcher.
+     *
+     * @param HookDispatcher $dispatcher The dispatcher.
+     */
+    public function setDispatcher(HookDispatcher $dispatcher);
```
Comment (stof) [inline src/Behat/MinkExtension/Context/MinkAwareContext.php]: this is impossible. Adding a new method in the interface is a BC break (and I don't want to go to MinkExtension 3.0 for now)

**Comment by stof** [comment]
This implementation is a bad idea IMO. Many requests will still happen without triggering the hook:

- form submissions
- clicks on links
- ``visit`` calls which are done using the session directly rather than visitPath
- ``reload``, ``forward`` and ``back``
- maybe other cases if you use a JS-capable driver, as the JS could perform some of these actions automatically when you do another one (for instance submitting a search form automatically when it is filled).

Providing a hook which runs for some requests but not all of them is way worse than not having it IMO.
And a hook running on all requests simply cannot be achieved from here, as it is too high level.

So it is a -1 from me.
Btw, I'm not a fan of injecting the hook dispatcher in all mink contexts either

If you actually need to run some logic for each requests, I see 2 solutions:

- identify all methods of a driver which can trigger a request (the list is given above for the case of non-JS-enabled drivers) and create a driver triggering some logic in them (hardcoded logic, or triggering an event, or triggering a hook if you really need to have contexts able to react to it but it would be weird) and delegating to the actual driver. Your own behat extension could then manipulate the container (in ``process``) to wrap drivers into your own driver.
- decide that you only care about the GoutteDriver, and hook at the guzzle level to do it (much easier as they already trigger event for requests, so you would only have to register a listener in the guzzle client by using a Behat extension to change the container)

**Comment by pfrenssen** [comment]
@stof thanks for the feedback.

I need this for the Behat Drupal Extension so it should work for each driver that Mink supports. I'll look into your first suggestion.

We have two distinct use cases at the moment: we need to be able to set a cookie before each request, and clear static caches inside our bootstrapped Drupal instance after each request. This can probably be done with event listeners instead of context hooks, so I'll look into that.

**Pull request closed without merging**